### PR TITLE
Bittree updates

### PR DESCRIPTION
**Pull request by akashdhruv**
## Summary

This PR introduces updates to regridding using [Bittree](https://github.com/Flash-X/Bittree)

## Additional background

We have introduced a new feature to use AMR in Octree mode. This was a result of ongoing performance studies which identified that Octree mode using Bittree scales better than AMReX native patch-based implementation.

## Checklist

The proposed changes:
- [ ] fix a bug or incorrect behavior in AMReX
- [x] add new capabilities to AMReX
- [ ] changes answers in the test suite to more than roundoff level
- [ ] are likely to significantly affect the results of downstream AMReX users
- [x] include documentation in the code and/or rst files, if appropriate

### CODE REVIEW

**Comment by akashdhruv** [comment]
Currently performing tests. This PR is created to enable discussion as we tweak the code and get it ready for production.

**Comment by rmrsk** [comment]
I'm curious what is meant by 'octree-mode' since octrees have a regular/leaf node separation where potentially nothing is allocated for regular nodes. But that's (almost) never the case for patch-based AMR...

**Comment by akashdhruv** [comment]
In the context of this PR octree mode refers to an octree refinement using the patch based algorithm. The idea is to be able to do regridding in this configuration using `Bittree` and bypass the native `MakeNewGrids`. Bittree has shown performance improvement for PARAMESH for Flash-X and we want to see if it can improve regridding for AMReX

**Comment by tklos96** [comment]
@WeiqunZhang I made various optimization changes and updated the passing of pointers as you suggested. I have one last problem before it's ready to merge - Subcycling. When I turn subcycling on for the Advection test, I run into issues after ~35 timesteps - the solution blows up and the grid layout becomes unpredictable. I am wondering if the Bittree logic doesn't properly account for the `lbase` or something. I can't tell what's happening though, any suggestions?

**Comment by WeiqunZhang** [comment]
We need to introduce `USE_BITTREE` into GNUMake system and add `-DAMREX_USE_BITTREE` to the compiler flags.  Then we can use `ifdef AMREX_USE_BITTREE` on the bittree code that has extra dependence.

File: Src/AmrCore/AMReX_Bittree.H [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+class btUnit {
+  // Functions used in AmrMesh
+  public:
+    static int  btRefine(BittreeAmr* const mesh,
+                         std::vector<int>& btTags,
+                         MPI_Comm comm);
+    static void btCalculateGrids(BittreeAmr* const mesh,
+                                 int lbase, Real time, int& new_finest,
+                                 Vector<BoxArray>& new_grids,
+                                 Vector<IntVect>& max_grid_size);
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Bittree.H]:
```suggestion
                                 Vector<IntVect> const& max_grid_size);
```

File: Src/AmrCore/AMReX_Bittree.H [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    static void btCalculateLevel(BittreeAmr* const mesh,
+                                 int lev, Real time,
+                                 BoxArray& ba,
+                                 IntVect& max_grid_size);
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Bittree.H]:
```suggestion
                                 IntVect const& max_grid_size);
```

File: Src/AmrCore/AMReX_Bittree.H [hunk 3]
```diff
@@ -0,0 +1,73 @@
+#ifndef BL_BITTREE_H_
+#define BL_BITTREE_H_
+
+#include <AMReX_IntVect.H>
+#include <AMReX_BoxArray.H>
+#include <AMReX_DistributionMapping.H>
+
+#include <Bittree_BittreeAmr.h>
+
+using namespace bittree;
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Bittree.H]: Can we avoid `using namespace` in header files?

**Review by WeiqunZhang — COMMENTED** [review]
Some comments.  I haven't finished reviewing yet.

File: Src/AmrCore/AMReX_Bittree.cpp [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+
+    BoxList bl;
+
+    for(int i=id0; i<id1; ++i) {
+      //Get coordinates and morton index.
+      auto b = tree1->locate(i);
+
+      if(b.level != lev) {
+          std::string msg = "Error identifying block in btCalculateGrids";
+          //throw error?
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Bittree.cpp]: `amrex::Abort("Error identify.....");`?

File: Src/AmrCore/AMReX_Bittree.cpp [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+    for(int lev=lbase; lev<=new_finest; ++lev) {
+        btCalculateLevel(mesh, lev, time, new_grids[lev],
+                         max_grid_size[lev]);
+    }
+}
+
+/** Creates a box array based on Bittree.
+  */
+void btUnit::btCalculateLevel(BittreeAmr* const mesh, int lev,
+                            Real time,
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_Bittree.cpp]:
```suggestion
                            Real /*time*/,
```
The CI is very strict on unused parameter.

File: Src/AmrCore/AMReX_AmrMesh.cpp [hunk 6]
```diff
@@ -762,6 +770,66 @@ AmrMesh::MakeNewGrids (int lbase, Real time, int& new_finest, Vector<BoxArray>&
...
+        // [1] Error Estimation and tagging
+        // btTags is indexed by bitid, Bittree's internal indexing scheme.
+        // For any id, btTags = 1 if should be parent, -1 if should not be parent (or not exist).
+        std::vector<int> btTags(tree0->id_upper_bound(),0);
+
+        for (int lev=max_crse; lev>=lbase; --lev) {
+
+            TagBoxArray tags(grids[lev],dmap[lev], n_error_buf[lev]);
+            ErrorEst(lev, tags, time, 0);
+            // ?? tags.buffer(n_error_buf[lev]);
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_AmrMesh.cpp]: Yes, we need this for error buffer.  And I think we need to do something similar to the ParallelAdd part (not the mask part) of `TagBoxArray::mapPeriodicRemoveDuplicates` to parallel communicate the buffer tags to valid cells.

File: Src/AmrCore/AMReX_AmrMesh.cpp [hunk 7]
```diff
@@ -762,6 +770,66 @@ AmrMesh::MakeNewGrids (int lbase, Real time, int& new_finest, Vector<BoxArray>&
...
+
+            for (MFIter mfi(tags); mfi.isValid(); ++mfi) {
+                auto const& tagbox = tags.const_array(mfi);
+                bool has_set_tags = amrex::Reduce::AnyOf(mfi.validbox(),
+                                                         [=] AMREX_GPU_DEVICE (int i, int j, int k)
+                                                         {
+                                                              return tagbox(i,j,k) == TagBox::SET;
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_AmrMesh.cpp]:
```suggestion
                                                              return tagbox(i,j,k) != TagBox::CLEAR;
```
to include buffer cells.

**Review by WeiqunZhang — COMMENTED** [review]
More comments

File: Tests/Amr/Advection_AmrCore/Exec/inputs [hunk 8]
```diff
@@ -2,7 +2,7 @@
 # Run until nsteps == max_step or time == stop_time, 
 #     whichever comes first
 # *****************************************************************
-max_step  = 1000
+max_step  = 100
```
Comment (WeiqunZhang) [inline Tests/Amr/Advection_AmrCore/Exec/inputs]: Can we keep the current inputs file and add a new file for the bittree test?

**Comment by WeiqunZhang** [comment]
For the failed subcycling test, is it 2d or 3d?

**Comment by tklos96** [comment]
It was 2d.

**Comment by akashdhruv** [comment]
> We need to introduce `USE_BITTREE` into GNUMake system and add `-DAMREX_USE_BITTREE` to the compiler flags. Then we can use `ifdef AMREX_USE_BITTREE` on the bittree code that has extra dependence.

This is done. Also created a separate directories `Src/Extern/Bittree` and `Tools/GNUmake/packages/Make.bittree` to manage source code and dependencies.

**Comment by akashdhruv** [comment]
`./configure --enable-bittree=yes` is the new option to build AMReX in Bittree mode. Default is `no`. Environment variable `BITTREE_$(DIM)D_HOME` should be defined to store Bittree library path. `$(DIM)` is the dimension

**Comment by akashdhruv** [comment]
Additional task: Setup a test in AMReX test suite @WeiqunZhang can you do that on your end?

File: Tests/Amr/Advection_AmrCore/Exec/inputs (hunk 8 above)
Comment (akashdhruv) [reply Tests/Amr/Advection_AmrCore/Exec/inputs]: Done

File: Tests/Amr/Advection_AmrCore/Exec/inputs [hunk 9]
```diff
@@ -31,15 +31,17 @@ amr.max_level       = 2       # maximum level number allowed --
 
 amr.ref_ratio       = 2 2 2 2 # refinement ratio between levels
 
+amr.use_bittree = true
+
 # *****************************************************************
 # Control of grid creation
 # *****************************************************************
 # Blocking factor for grid creation in each dimension --
 #   this ensures that every grid is coarsenable by a factor of 8 --
```
Comment (kweide) [inline Tests/Amr/Advection_AmrCore/Exec/inputs]: Does the comment ("every grid is coarsenable by a factor of 8") need updating?

**Comment by WeiqunZhang** [comment]
Yes, once the bittree repo is public, I will set up a regression test.

File: Tests/Amr/Advection_AmrCore/Exec/inputs (hunk 9 above)
Comment (WeiqunZhang) [reply Tests/Amr/Advection_AmrCore/Exec/inputs]: Good question. @akashdhruv and @tklos96 Do we still need to set blocking factors and max_grid_size in `inputs_bittree`?

File: Src/AmrCore/AMReX_AmrMesh.H [hunk 10]
```diff
@@ -260,6 +264,11 @@ protected:
     Vector<DistributionMapping> dmap;
     Vector<BoxArray>            grids;
 
+#ifdef AMREX_USE_BITTREE
+    bool use_bittree;
```
Comment (WeiqunZhang) [inline Src/AmrCore/AMReX_AmrMesh.H:268]: I think we should move this flag into `AmrInfo` and gives it a default value of false.  One constructor of AmrMesh does not use ParmParse at all and it relies on the parameters in `AmrInfo`.  Then for consistence, in `InitAmrMesh`, we should change from `get` to `queryAdd`.

**Comment by WeiqunZhang** [comment]
The last time I tried, I could not reproduce the blow-up issue with subcycling.  After the two minor comments are addressed, I think we can merge this first, because this is a big PR.

**Comment by akashdhruv** [comment]
I found an issue with a 3D setup in Flash-X. Give me a few days to resolve that and then we can merge

**Comment by akashdhruv** [comment]
Performance of Paramesh vs AMReX in Flash-X

bittree

**Comment by akashdhruv** [comment]
Continued in #3547

**Pull request closed without merging**